Implemented `._partition_args()` function for Rubi pattern matcher

arihantparsoya opened the pull request:
I have implemented `._partition_args()` function which can be used to make new Rubi pattern matcher efficient. Pattern matching of `is_commutative` expressions(`Add` and `Mul`) leads to exponential time. By partitioning the expression into subexpressions which contain integration variable and which does not, we can reduce the search space of matcher.

`expr.partition(var)` function returns a `dict` of subexpressions which contains `var`. Subexpressions which doesn't contain `var` are considered to be `Constant`.

#### Sample Program:
``` python
>>> from sympy import exp, sqrt, cos, sin
>>> from sympy.abc import x, a, b
>>> expr = a + b + x
>>> expr._partition_args(x)
{'Constant': [a, b], 'Symbol': [x]}
>>> expr = cos(a)*b*sin(x)
>>> expr._partition_args(x)
{'Constant': [b, cos(a)], 'sin': [sin(x)]}
```

Reference:

[https://people.eecs.berkeley.edu/~fateman/papers/partition-new2.pdf](https://people.eecs.berkeley.edu/~fateman/papers/partition-new2.pdf)


* Tests are added.

arihantparsoya committed:
Implemented `_partition_args()` Function

Reference:

https://people.eecs.berkeley.edu/~fateman/papers/partition-new2.pdf

 Partition function returns a `dict` of subexpressions which contains
`var`. Subexpressions which does't contain `var` are considered to be
constant.

Upabjojr commented on sympy/core/basic.py:
```diff
@@ -1525,6 +1525,52 @@ def match(self, pattern, old=False):
         pattern = sympify(pattern)
         return pattern.matches(self, old=old)
 
+    def partition(self, var):
```
Maybe we should make it private, e.g. by calling it `_partition`. Maybe `_partition_args` could be a more suitable name. Is it intended for end-users?

parsoyaarihant replied:
Thanks! This function is not for end-users.

I have renamed the function to `_partiton_args()`

Upabjojr commented on sympy/core/basic.py:
```diff
@@ -1525,6 +1525,52 @@ def match(self, pattern, old=False):
+        leaves = {}
+        leaves['Constant'] = []
+
+        if self.__class__.__name__  not in ['Mul', 'Add']:
```
```python
from .add import Add
from .mul import Mul
if not isinstance(self, [Mul, Add]):
```

This would care for a possible inheritance of Mul and Add.

parsoyaarihant commented:
@certik please have a look and let me know of improvements.

Upabjojr committed:
Added reference to paper

parsoyaarihant commented on sympy/core/basic.py:
```diff
@@ -1530,6 +1530,9 @@ def _partition_args(self, var):
         Subexpressions which doesn't contain `var` are considered to be constant.
 
+        See:
+        https://people.eecs.berkeley.edu/~fateman/papers/partition-new2.pdf
```
@Upabjojr , thanks.

smichr commented on sympy/core/basic.py:
```diff
@@ -1525,6 +1525,57 @@ def match(self, pattern, old=False):
+        leaves['Constant'] = []
+
+        if not isinstance(self, (Mul, Add)):
+            if self.has(var):
```
-> `if var in self.free_symbols`

(and the same change throughout because `has` does not imply dependence on a given symbol)

```
>>> Integral(x, (x, 1, 0)).has(x)
True
>>> Integral(x, (x, 1, 0)).free_symbols
{}
```

smichr commented on sympy/core/basic.py:
```diff
@@ -1525,6 +1525,57 @@ def match(self, pattern, old=False):
+
+        for subexpr in self.args:
+            if subexpr.has(var):
+                try:
```
replace try/except with

```
leaves.setdefault(subexpr.__class__.__name__, []).append(subexpr)
```

parsoyaarihant replied:
Thanks, I have committed the changes.

arihantparsoya committed:
Removed exceptions

Added changes suggested by smichr.

smichr commented on sympy/core/basic.py:
```diff
@@ -1525,6 +1525,54 @@ def match(self, pattern, old=False):
+        {'Constant': [b, cos(a)], 'sin': [sin(x)]}
+
+        When the expression head is not commutative('Add' and 'Mul' type
+        epressions), the same expression is returned in form of dict.
```
`Expr.args_cnc` may be of use to you.

Upabjojr replied:
Easiest way: raise an error if non-commutative terms are encountered.

parsoyaarihant replied:
Can we have another function parameter which "raise a `NotImplementedError`" or "treat all terms as commutative" when non-commutative terms are encountered?

Upabjojr replied:
`NotImplementedError` can be fine, as probably in the future we should try to generalize it.

parsoyaarihant replied:
added it.

arihantparsoya committed:
raises NotImplementedError when non commutative terms are encountered

Defined new function variable `all_comm`. When `all_comm` is set to
True, the function treats all terms in the expression as commutative.

If set to false, the function raises NotImplementedError for now.

czgdp1807 commented:
@Upabjojr @smichr @parsoyaarihant Please update the status of this PR.

parsoyaarihant commented:
I believe this PR can be closed. This work was addressed in the GSoC '17 RUBI project.
